### PR TITLE
Adding host to the lobby.

### PR DESCRIPTION
**Pull request by YingLiang2**
The lobbies now have a host indicated by a crown symbol next to their name (ICON can be changed later)

When a new lobby is created, the user who created the host will become the host of that lobby. 

Every subsequent player that joins will be `waiting for the host...` to start the game.

When the host leaves the game, if there are any players in that lobby then the game will randomly assign one of them to become the host. 

image
image

The width of these lobbies is not the same because the `waiting for host...` text is bigger than `Start Game` so we might we want to figure that out later.

### CODE REVIEW

File: server/src/room.ts [hunk 1]
```diff
@@ -45,5 +45,28 @@ export default class Room {
 
     isEmpty() {
         return this.#players.size === 0;
+    }
+
+    setLeader(playerID: string) {
```
Comment (YingLiang2) [inline server/src/room.ts:50]: Will make another commit, removing all the extra space.

File: web/src/app/components/Game.tsx [hunk 2]
```diff
@@ -6,6 +6,7 @@ import socket from '@/scripts/SocketConnection';
 import ButtonSocketConnection from "./ButtonSocketConnection";
 import PlayerState from "../interfaces/PlayerState";
 import Lobby from "./Lobby";
+import PreviousMap from "postcss/lib/previous-map";
```
Comment (YingLiang2) [inline web/src/app/components/Game.tsx:9]: Will remove this. This wasn't used at all and it was a mistake from an import.

File: web/src/app/components/LobbyPlayerList.tsx [hunk 3]
```diff
@@ -6,12 +6,22 @@ interface LobbyPlayerListProps {
...
+        <div className="inline-block overflow-auto rounded-lg bg-[#EAEAEA] relative px-4 py-2"> 
             <h2 className="text-black">Connected Players ({Object.keys(players).length}/4)</h2>
             <ul>
-                {Object.entries(players).map(([id, {username}], index) => (
-                    <div key={id + username} className="mt-1">
-                         <li className="subtext text-black">{username}</li>  
+                {Object.entries(players).map(([id, {username, host}], index) => (
+                    <div key={id + username} className={`mt-1 ${host ? 'flex justify-items-center' : '' }`}>
+                        
+                        {host &&  <svg
```
Comment (YingLiang2) [inline web/src/app/components/LobbyPlayerList.tsx:15]: I think we probably will need a new icon for indicating the host, but this one can do for now I suppose.

File: web/src/app/multiplayer/page.tsx [hunk 4]
```diff
@@ -80,38 +80,32 @@ export default function Multiplayer() {
...
-                            <Table.Thead>
-                                <Table.Tr>
-                                    <Table.Th className="pl-0">Code</Table.Th>
-                                    <Table.Th className="pl-14">Players</Table.Th>
-                                    <Table.Th className="pl-14 invisible" >Players</Table.Th>  {/* Stupid hack to make the header fixed in position when the filter returns 0 results*/}
-                                </Table.Tr>
-                            </Table.Thead>
-                           <Table.Tbody>{rows}</Table.Tbody>
-                        </Table>
+        <div className='flex justify-center items-center my-10'>
```
Comment (YingLiang2) [inline web/src/app/multiplayer/page.tsx:83]: Tried to refactor the multiplayer list css

File: web/src/app/components/Lobby.tsx [hunk 5]
```diff
@@ -1,19 +1,21 @@
-import React, { useRef, useState } from 'react';
+import React, { useEffect, useRef, useState } from 'react';
```
Comment (YingLiang2) [inline web/src/app/components/Lobby.tsx:1]: So for the logic of finding out who is host is done in this Lobby component but it easily could have been done in the `Game.tsx` component as well, and we could have just passed a boolean prop into this `Lobby` component. I'm not sure which one is better so what are your thoughts?

File: web/src/app/components/Lobby.tsx [hunk 6]
```diff
@@ -30,6 +32,14 @@ export default function Lobby({roomID, players, onStart, onLeave}: LobbyProps) {
         
     }
 
+    useEffect(() => {
```
Comment (YingLiang2) [inline web/src/app/components/Lobby.tsx]: I'm also not sure if this is the best way of finding the host since this useEffect code will run everytime the players list gets updated, which can be often.

File: web/src/app/components/Lobby.tsx (hunk 5 above)
Comment (jonathanplacatka) [reply web/src/app/components/Lobby.tsx:1]: I think putting it in Game.tsx is better. instead of having this useEffect, we can just put the check for host inside `socket.on('allPlayers')`. That way we can also be sure the host check is only run when a player joins/leaves (addressing the useEffect problem you mentioned in the other comment)

something like:

```
        socket.on('allPlayers', (players) => {
            setPlayers(players);
            setIsCurrPlayerHost(players[currPlayerID].host);
        })
```

I think it's better to handle the player join/leaving logic in one place as well

**Comment by jonathanplacatka** [comment]
can we move the host icon up (so it's centered vertically with the text), and to the left a bit?

**Comment by jonathanplacatka** [comment]
can we set a min-width on the buttons so they don't change size depending on the text? just add:

`className="min-w-[150px]"`

(adjusting the px until they look the same)

File: web/src/app/components/Lobby.tsx (hunk 5 above)
Comment (YingLiang2) [reply web/src/app/components/Lobby.tsx:1]: Yeah I agree, but I think I have to useEffect() here because of react's state changes. `currPlayerID` is always empty so I can't do anything in that socketevent

Comment (jonathanplacatka) [reply web/src/app/components/Lobby.tsx:1]: currPlayerId should be set on the connect event. I think this will work

**Comment by YingLiang2** [comment]
image
image

Changed button styling to match now.

File: web/src/app/components/Lobby.tsx (hunk 5 above)
Comment (YingLiang2) [reply web/src/app/components/Lobby.tsx:1]: I tried and it didn't work. It's the first thing I tried.

**Comment by YingLiang2** [comment]
Can you also pull my branch and test it out to see if there any bugs @jonathanplacatka?

File: web/src/app/components/Lobby.tsx (hunk 5 above)
Comment (jonathanplacatka) [reply web/src/app/components/Lobby.tsx:1]: you're right, since currPlayerID uses useState, it's not updated until the next render. We can just use `useRef `instead.

useRef is similar to state except it doesn't trigger a re-render when the value is changed. But since currPlayerID just used internally and not for any UI, useRef is better anyway

**Review by jonathanplacatka — APPROVED** [review]
